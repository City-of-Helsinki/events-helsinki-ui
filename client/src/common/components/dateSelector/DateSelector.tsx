import React, { FunctionComponent } from "react";

import { ReactComponent as AngleDownIcon } from "../../../assets/icons/svg/angle-down.svg";
import { ReactComponent as AngleUpIcon } from "../../../assets/icons/svg/angle-up.svg";
import { ReactComponent as CalendarIcon } from "../../../assets/icons/svg/calendar.svg";
import { formatMessage } from "../../translation/utils";
import styles from "./dateSelector.module.scss";
import DateSelectorMenu from "./DateSelectorMenu";

interface Props {
  dateTypes: string[];
  endDate: Date | null;
  isCustomDate: boolean;
  onChangeDateTypes: (value: string[]) => void;
  onChangeEndDate: (date: Date | null) => void;
  onChangeStartDate: (date: Date | null) => void;
  startDate: Date | null;
  toggleIsCustomDate: () => void;
}

const DateSelector: FunctionComponent<Props> = ({
  dateTypes,
  endDate,
  isCustomDate,
  onChangeDateTypes,
  onChangeEndDate,
  onChangeStartDate,
  startDate,
  toggleIsCustomDate
}) => {
  const backBtnRef = React.useRef<HTMLButtonElement | null>(null);
  const toggleBtnRef = React.useRef<HTMLButtonElement | null>(null);
  const dateSelector = React.useRef<HTMLDivElement | null>(null);
  const [isMenuOpen, setIsMenuOpen] = React.useState(false);

  const handleDocumentClick = (event: MouseEvent) => {
    const target = event.target;
    const current = dateSelector && dateSelector.current;

    // Close menu when clicking outside of the component
    if (!(current && target instanceof Node && current.contains(target))) {
      setIsMenuOpen(false);
    }
  };

  const handleDocumentKeyDown = (event: KeyboardEvent) => {
    switch (event.keyCode) {
      // Close menu on ESC key
      case 27:
        setIsMenuOpen(false);
        break;
    }
  };

  const handleDocumentFocusin = (event: FocusEvent) => {
    const target = event.target;
    const current = dateSelector && dateSelector.current;

    if (!(current && target instanceof Node && current.contains(target))) {
      setIsMenuOpen(false);
    }
  };

  const toggleMenu = () => {
    setIsMenuOpen(!isMenuOpen);
  };

  React.useEffect(() => {
    document.addEventListener("click", handleDocumentClick);
    document.addEventListener("keydown", handleDocumentKeyDown);
    document.addEventListener("focusin", handleDocumentFocusin);
    // Clean up event listener to prevent memory leaks
    return () => {
      document.removeEventListener("click", handleDocumentClick);
      document.removeEventListener("keydown", handleDocumentKeyDown);
      document.removeEventListener("focusin", handleDocumentFocusin);
    };
  }, []);

  const handleToggleIsCustomDate = () => {
    if (isCustomDate) {
      setTimeout(() => {
        if (toggleBtnRef && toggleBtnRef.current) {
          toggleBtnRef.current.focus();
        }
      }, 1);
    } else {
      setTimeout(() => {
        if (backBtnRef && backBtnRef.current) {
          backBtnRef.current.focus();
        }
      }, 1);
    }

    toggleIsCustomDate();
  };

  return (
    <div className={styles.dateSelector} ref={dateSelector}>
      <button className={styles.button} onClick={toggleMenu} type="button">
        <div className={styles.iconWrapper}>
          <CalendarIcon />
        </div>
        <div className={styles.info}>
          {formatMessage("commons.dateSelector.title")}
        </div>
        <div className={styles.arrowWrapper}>
          {isMenuOpen ? <AngleUpIcon /> : <AngleDownIcon />}
        </div>
      </button>
      <DateSelectorMenu
        backBtnRef={backBtnRef}
        dateTypes={dateTypes}
        endDate={endDate}
        isCustomDate={isCustomDate}
        isOpen={isMenuOpen}
        onChangeDateTypes={onChangeDateTypes}
        onChangeEndDate={onChangeEndDate}
        onChangeStartDate={onChangeStartDate}
        startDate={startDate}
        toggleBtnRef={toggleBtnRef}
        toggleIsCustomDate={handleToggleIsCustomDate}
        toggleMenu={toggleMenu}
      />
    </div>
  );
};

export default DateSelector;
