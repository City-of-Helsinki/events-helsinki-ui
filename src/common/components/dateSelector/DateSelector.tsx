import React, { FunctionComponent } from "react";
import { useTranslation } from "react-i18next";

import IconAngleDown from "../../../icons/IconAngleDown";
import IconAngleUp from "../../../icons/IconAngleUp";
import IconCalendar from "../../../icons/IconCalendar";
import styles from "./dateSelector.module.scss";
import DateSelectorMenu from "./DateSelectorMenu";

interface Props {
  dateTypes: string[];
  endDate: Date | null;
  isCustomDate: boolean;
  name: string;
  onChangeDateTypes: (value: string[]) => void;
  onChangeEndDate: (date: Date | null) => void;
  onChangeStartDate: (date: Date | null) => void;
  onSubmit?: () => void;
  startDate: Date | null;
  toggleIsCustomDate: () => void;
}

const DateSelector: FunctionComponent<Props> = ({
  dateTypes,
  endDate,
  isCustomDate,
  name,
  onChangeDateTypes,
  onChangeEndDate,
  onChangeStartDate,
  onSubmit,
  startDate,
  toggleIsCustomDate
}) => {
  const { t } = useTranslation();
  const backBtnRef = React.useRef<HTMLButtonElement | null>(null);
  const toggleBtnRef = React.useRef<HTMLButtonElement | null>(null);
  const dateSelector = React.useRef<HTMLDivElement | null>(null);
  const [isMenuOpen, setIsMenuOpen] = React.useState(false);

  const handleDocumentClick = (event: MouseEvent) => {
    const target = event.target;
    const current = dateSelector && dateSelector.current;

    // Close menu when clicking outside of the component
    if (!(current && target instanceof Node && current.contains(target))) {
      setIsMenuOpen(false);
    }
  };

  const shouldSubmit = () => {
    const active = document.activeElement;
    if (
      backBtnRef.current &&
      active instanceof Node &&
      backBtnRef.current.contains(active)
    ) {
      return false;
    }
    if (
      toggleBtnRef.current &&
      active instanceof Node &&
      toggleBtnRef.current.contains(active)
    ) {
      return false;
    }
    return true;
  };

  const ensureMenuIsOpen = React.useCallback(() => {
    if (!isMenuOpen) {
      setIsMenuOpen(true);
    }
  }, [isMenuOpen]);

  const isComponentFocused = React.useCallback(() => {
    const active = document.activeElement;
    const current = dateSelector && dateSelector.current;

    if (current && active instanceof Node && current.contains(active)) {
      return true;
    }
    return false;
  }, [dateSelector]);

  const handleSubmit = React.useCallback(() => {
    if (onSubmit) {
      onSubmit();
    }
    closeMenu();
  }, [onSubmit]);

  const handleDocumentFocusin = React.useCallback(() => {
    if (isComponentFocused()) {
      setIsMenuOpen(true);
    } else {
      setIsMenuOpen(false);
    }
  }, [isComponentFocused]);

  const handleDocumentKeyDown = React.useCallback(
    (event: KeyboardEvent) => {
      if (!isComponentFocused()) return;

      switch (event.key) {
        case "ArrowUp":
          ensureMenuIsOpen();
          event.preventDefault();
          break;
        case "ArrowDown":
          ensureMenuIsOpen();
          event.preventDefault();
          break;
        case "Escape":
          setIsMenuOpen(false);
          event.preventDefault();
          break;
        case "Enter":
          if (shouldSubmit()) {
            handleSubmit();
            event.preventDefault();
          }

          break;
      }
    },
    [ensureMenuIsOpen, handleSubmit, isComponentFocused]
  );

  const closeMenu = () => {
    setIsMenuOpen(false);
  };

  React.useEffect(() => {
    document.addEventListener("click", handleDocumentClick);
    document.addEventListener("keydown", handleDocumentKeyDown);
    document.addEventListener("focusin", handleDocumentFocusin);
    // Clean up event listener to prevent memory leaks
    return () => {
      document.removeEventListener("click", handleDocumentClick);
      document.removeEventListener("keydown", handleDocumentKeyDown);
      document.removeEventListener("focusin", handleDocumentFocusin);
    };
  }, [handleDocumentFocusin, handleDocumentKeyDown]);

  const handleToggleIsCustomDate = () => {
    if (isCustomDate) {
      setTimeout(() => {
        if (toggleBtnRef && toggleBtnRef.current) {
          toggleBtnRef.current.focus();
        }
      }, 1);
    } else {
      setTimeout(() => {
        if (backBtnRef && backBtnRef.current) {
          backBtnRef.current.focus();
        }
      }, 1);
    }

    toggleIsCustomDate();
  };

  const isSelected = isCustomDate
    ? !!startDate || !!endDate
    : !!dateTypes.length;

  return (
    <div className={styles.dateSelector} ref={dateSelector}>
      <button
        className={styles.button}
        onClick={ensureMenuIsOpen}
        type="button"
      >
        <div className={styles.iconWrapper}>
          <IconCalendar />
        </div>
        <div className={styles.info}>{t("commons.dateSelector.title")}</div>
        <div className={styles.arrowWrapper}>
          {isMenuOpen ? <IconAngleUp /> : <IconAngleDown />}
        </div>
      </button>
      {isSelected && <div className={styles.isSelectedIndicator} />}
      <DateSelectorMenu
        backBtnRef={backBtnRef}
        dateTypes={dateTypes}
        endDate={endDate}
        isCustomDate={isCustomDate}
        isOpen={isMenuOpen}
        name={name}
        onChangeDateTypes={onChangeDateTypes}
        onChangeEndDate={onChangeEndDate}
        onChangeStartDate={onChangeStartDate}
        startDate={startDate}
        toggleBtnRef={toggleBtnRef}
        toggleIsCustomDate={handleToggleIsCustomDate}
        onCloseMenu={closeMenu}
      />
    </div>
  );
};

export default DateSelector;
