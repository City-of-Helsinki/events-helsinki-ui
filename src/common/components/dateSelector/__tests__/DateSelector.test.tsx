import * as React from 'react';
import { MemoryRouter } from 'react-router';
import renderer from 'react-test-renderer';

import DateSelector from '../DateSelector';

test('DateSelector matches snapshot', () => {
  const component = renderer.create(
    <MemoryRouter>
      <DateSelector
        dateTypes={['type1', 'type2']}
        endDate={new Date('2019-12-01')}
        isCustomDate={false}
        name="date"
        onChangeDateTypes={() => {}}
        onChangeEndDate={() => {}}
        onChangeStartDate={() => {}}
        startDate={new Date('2019-11-01')}
        toggleIsCustomDate={() => {}}
      />
    </MemoryRouter>
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

export {};
