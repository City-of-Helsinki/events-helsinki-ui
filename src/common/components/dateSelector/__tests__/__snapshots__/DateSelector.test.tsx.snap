// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateSelector matches snapshot 1`] = `
<div
  className="dateSelector"
>
  <button
    aria-expanded={false}
    aria-haspopup="true"
    className="button"
    onClick={[Function]}
    type="button"
  >
    <div
      className="iconWrapper"
    >
      <svg
        className="iconCalendar"
        height="100%"
        viewBox="0 0 27 26"
        width="100%"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g
          fill="none"
          fillRule="evenodd"
          strokeWidth="1"
        >
          <g
            strokeWidth="2"
            transform="translate(-785.000000, -171.000000)"
          >
            <g
              transform="translate(773.000000, 129.000000)"
            >
              <g
                transform="translate(13.000000, 43.000000)"
              >
                <circle
                  cx="17.6"
                  cy="18.4"
                  r="5.6"
                  strokeLinecap="square"
                />
                <polyline
                  points="17.6 16 17.6 18.4 20 18.4"
                  strokeLinecap="square"
                />
                <line
                  strokeLinecap="square"
                  x1="6.4"
                  x2="6.4"
                  y1="0"
                  y2="3.2"
                />
                <line
                  strokeLinecap="square"
                  x1="17.6"
                  x2="17.6"
                  y1="0"
                  y2="3.2"
                />
                <line
                  x1="24"
                  x2="0"
                  y1="8"
                  y2="8"
                />
                <polyline
                  points="8.8 22.4 0 22.4 0 3.2 24 3.2 24 11.2"
                  strokeLinecap="square"
                />
              </g>
            </g>
          </g>
        </g>
      </svg>
    </div>
    <div
      className="info"
    >
      Valitse ajankohta
    </div>
    <div
      className="arrowWrapper"
    >
      <svg
        className=""
        height="100%"
        viewBox="0 0 199.404 199.404"
        width="100%"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g>
          <polygon
            points="199.404,63.993 171.12,35.709 99.702,107.127 28.284,35.709 0,63.993 99.702,163.695"
          />
        </g>
      </svg>
    </div>
  </button>
  <div
    className="isSelectedIndicator"
  />
</div>
`;
